File: 238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, 1);
        int left = 1, right = 1;
        for (int i = 0, j = nums.size() - 1; i < nums.size() - 1; i++, j--) {
            left *= nums[i];
            right *= nums[j];
            res[i + 1] *= left;
            res[j - 1] *= right;
        }
        return res;
    }
};

File: java/238. Product of Array Except Self.java
// O(n)
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        int accum = 1;
        for (int i = 0; i < nums.length; ++i) {
            res[i] = accum;
            accum *= nums[i];
        }
        accum = 1;
        for (int i = nums.length - 1; i >= 0; --i) {
            res[i] *= accum;
            accum *= nums[i];
        }
        return res;
    }
}
